### PR TITLE
Call optimizer outside the loop in compute gradient on summed gradients

### PR DESCRIPTION
**Pull request by akorgor**
This PR aims to speed up the code by calling the `optimized_weight` function outside the loop in `compute_gradient`.

The following plots were obtained from running the `eprop_supervised_classification_evidence-accumulation.py` task (according to the new file naming, i.e., with the non `_bsshslm_2020` models).

For a cutoff of 10, there is a small speed-up while the losses are accurate:

image
image

For a cutoff of 1000, there is a significant speed-up whilst still maintaining accurate losses:

image
image
image

### CODE REVIEW

**Comment by JesusEV** [comment]
@akorgor This is a strong optimization. 

It would be even better not to pass the optimizer to `compute gradient` and perform the weight update from the `send` function of the synapse.

There is a small point of contention regarding the Adam optimizer.

Given that Adam maintains individual state variables for each parameter, updated according to the cumulative number of computations, the update frequencies of these variables may vary among synapses. Some synapses may exhibit higher firing rates, leading to more frequent updates of their state variables, while others may fire less frequently, resulting in less frequent updates.

I'm uncertain whether this aligns with the intended use of the Adam optimizer.

If this discrepancy in update frequencies among synapses does not lead to suboptimal convergence behavior or slow training I think we can consider this optimization for the case of Adam as well.

**Comment by akorgor** [comment]
Running simulations of `eprop_supervised_classification_evidence-accumulation.py` with
- the corrected loss
- a cutoff of 1000
- the optimizer called from the send function of the synapse and not passed to the `compute_gradient` function

show a significant speed-up but with a worse prediction error. However, even the baseline training prediction error is not good yet, so we need to tweak the parameters.

image
image
image
image

**Comment by akorgor** [comment]
- [x] The problem with this implementation is that the `optimize_each_step` flag is irrelevant for the `_bsshslm_2020` model, but is still around in the weight optimizer dictionary. Is that a problem? Maybe remove from the tutorial dictionaries?
- [x] Add somewhere a more detailed explanation of the two options with up- and downsides.

**Comment by JesusEV** [comment]
> * [ ]  The problem with this implementation is that the `optimize_each_step` flag is irrelevant for the `_bsshslm_2020` model, but is still around in the weight optimizer dictionary. Is that a problem? Maybe remove from the tutorial dictionaries?
> * [ ]  Add somewhere a more detailed explanation of the two options with up- and downsides.

@akorgor Can the bsshslm_2020 models be modified to throw an error with the message, "This model does not support optimization at each step" or similar?

Perhaps this could be done from the `check_connection` function in `eprop_synapse`? (there is access to the optimizer there)

File: models/weight_optimizer.h [hunk 1]
```diff
@@ -68,21 +68,29 @@ The TensorFlow implementation deviates from [1]_ in that it assumes
 :math:`\hat{\epsilon} = \epsilon \sqrt{ 1 - \beta_2^t }` to be constant, whereas [1]_
 assumes :math:`\epsilon = \hat{\epsilon} \sqrt{ 1 - \beta_2^t }` to be constant.
 
+When `optimize_each_step` is set to `True`, the weights are optimized at every
+time step. If set to `False`, optimization occurs once per spike, resulting in a
+significant speed-up. For gradient descent, both settings yield the same
+results. However, for the Adam optimizer, only setting `optimize_each_step` to
+`True` precisely implements the algorithm as described in [2]_. The impact of
+this setting on learning performance may vary depending on the task.
```
Comment (JesusEV) [inline models/weight_optimizer.h]:
```suggestion
results under exact arithmetic; however, small numerical differences may be 
observed due to floating point precision. For the Adam optimizer, only setting
`optimize_each_step` to `True` precisely implements the algorithm as described
in [2]_. The impact of this setting on learning performance may vary depending
on the task.
```
I think this clarification could avoid confusion in case a user starts observing small deviations after several iterations when comparing the 2 options for gradient descent.

**Review by JesusEV — APPROVED** [review]

File: models/weight_optimizer.h (hunk 1 above)
Comment (akorgor) [reply models/weight_optimizer.h]: That's a good point! Is there a word missing after arithmetic?

Comment (JesusEV) [reply models/weight_optimizer.h]: @akorgor no, I don't think so as long as "exact arithmetic" is interpreted as "without any loss of precision or rounding errors"

Comment (akorgor) [reply models/weight_optimizer.h]: Ah, sorry, I thought it was an adjective and the corresponding noun would be arithmetics - but you are right, the noun is arithmetic.
